app/controllers/api/v2/hosts_controller.rb

```diff
@@ -163,6 +163,23 @@ def update
         render_error :custom_error, :status => :unprocessable_entity, :locals => { :message => e.to_s }
       end
 
+      api :POST, "/hosts/:id/clone", N_("Clone a host")
+      param :id, :identifier_dottable, :required => true, :desc => N_("ID of host to be cloned")
+      param :name, String, :required => true, :desc => N_("New name of cloned host")
+      param :mac, String, :required => false, :desc => N_("MAC address of cloned host")
+      param :ip, String, :required => true, :desc => N_("IP address of cloned host")
+      param_group :host, :as => clone
+
+      def clone
+        @cloned_host = @host.api_clone
```

Comment:
I Tried to create a host with wrong IP address (for example the IP isn't in the correct  subnet)
and I got the following error: 
Could not clone the host:
  resource have no errors

I've noticed that when updating the host (instead of clone) I'm getting the correct error: 
Ip does not match selected subnet

After some investigations, I saw it's related to the function process_response that calls process_response function and it calls to process_resource_error which eventually calls get_resource function. 
and get_resource returns @host resource and not @cloned_host.

Therefore, my suggestion is to change  @cloned_host to @host: 
      def clone
        @host = @host.api_clone
        @host.build = true if @host.managed?
        @host.attributes = host_attributes(host_params, @host)
        process_response @host.save
      rescue InterfaceTypeMapper::UnknownTypeExeption => e
        render_error :custom_error, :status => :unprocessable_entity, :locals => { :message => e.to_s }
      end

Reply:
@shiramax Thanks so much for advice. Changed. Now returning right error message. Pr will be ready soon but I need to solve problem with overridden puppetclasses as Oprazak mentioned above.